frontend/rust-lib/flowy-search/src/folder/indexer.rs

```diff
@@ -0,0 +1,223 @@
+use std::{any::Any, collections::HashMap, fs, path::Path, sync::Weak};
+
+use crate::folder::schema::{FolderSchema, FOLDER_TITLE_FIELD_NAME};
+use collab::core::collab::{IndexContent, IndexContentReceiver};
+use collab_folder::ViewIndexContent;
+use flowy_error::FlowyError;
+use flowy_user::services::authenticate_user::AuthenticateUser;
+use lib_dispatch::prelude::af_spawn;
+use tantivy::{
+  collector::TopDocs, directory::MmapDirectory, doc, query::QueryParser, DocAddress, Index, Score,
+  Term,
+};
+
+use crate::{
+  entities::SearchResultPB,
+  services::indexer::{IndexManager, IndexableData},
+};
+
+use super::{entities::FolderIndexData, schema::FOLDER_ID_FIELD_NAME};
+
+#[derive(Clone)]
+pub struct FolderIndexManager {
+  folder_schema: FolderSchema,
+  index: Index,
+}
+
+const FOLDER_INDEX_DIR: &str = "folder_index";
+
+impl FolderIndexManager {
+  pub fn new(auth_user: Weak<AuthenticateUser>) -> Self {
+    let authenticate_user = auth_user.upgrade();
+    let storage_path = match authenticate_user {
+      Some(authenticate_user) => authenticate_user.get_index_path(),
+      None => {
+        panic!("The user is not available");
+      },
+    };
+
+    let index_path = storage_path.join(Path::new(FOLDER_INDEX_DIR));
+
+    if !index_path.exists() {
+      let _ = fs::create_dir_all(&index_path);
+    }
+
+    let dir = MmapDirectory::open(index_path);
+    let folder_schema = FolderSchema::new();
+
+    let dir = match dir {
+      Ok(dir) => dir,
+      Err(err) => {
+        panic!("Failed to open folder index directory: {:?}", err)
+      },
+    };
+
+    let index = Index::open_or_create(dir, folder_schema.clone().schema).unwrap();
+
+    Self {
+      folder_schema,
+      index,
+    }
+  }
+
+  pub fn search(&self, query: String) -> Result<Vec<SearchResultPB>, FlowyError> {
+    let title_field = self
+      .folder_schema
+      .schema
+      .get_field(FOLDER_TITLE_FIELD_NAME)
+      .unwrap();
+
+    let length = query.len();
+    let distance: u8 = if length > 4 {
+      2
+    } else if length > 2 {
+      1
+    } else {
+      0
+    };
+
+    let mut query_parser = QueryParser::for_index(&self.index.clone(), vec![title_field]);
+    query_parser.set_field_fuzzy(title_field, true, distance, true);
+    let built_query = query_parser.parse_query(&query.clone()).unwrap();
+
+    let reader = self.index.reader().unwrap();
```

Comment:
If it's ok to unwrap, then adding documentation for it would be better. Make sure all the unwrap is safety